### ISSUE TITLE
Different behavior between MMSeg's cross_entropy loss and torch.nn.functional.cross_entropy when ignore_index is provided?

### ISSUE THREAD

**Issue by Wuziyi616** [open-mmlab/mmsegmentation#546]
Hi, currently I am training a segmentor on a dataset that contains several don't care classes. I set those class as `ignore_index` but discover that the loss value using your `cross_entropy` loss is different from `torch.nn.functional.cross_entropy`. According to [here](https://github.com/pytorch/pytorch/blob/master/torch/nn/functional.py#L2772), the `'mean'` reduction of loss is over non-ignored input (which is `unreduced_loss.sum() / (non_ignored_input.numel())`).

However, in MMSeg's implementation, since you split loss calculation and reduction into two steps, the reduction won't be aware of the ignored_input. And you calculate the reduction as `unreduced_loss.sum() / (all_input.numel())`. As you can see [here](https://github.com/open-mmlab/mmsegmentation/blob/master/mmseg/models/losses/cross_entropy_loss.py#L29-L30), the reduction step doesn't take label into consideration.

I think a solution is to calculate the `avg_factor` as `label.numel() - (label == ignore_index).sum().item()`, so that the reduced loss will be re-scaled to ignore those don't care input [here](https://github.com/open-mmlab/mmsegmentation/blob/master/mmseg/models/losses/utils.py#L71). If you think it's acceptable, I will create a PR for this?

**Comment by Junjun2016** [comment]
Hi @Wuziyi616 
Sorry for the late reply.
Looking forward to your PR！
Thanks so much！
